src/csharp/Microsoft.Spark.UnitTest/CallbackTests.cs

```diff
@@ -153,17 +153,33 @@ public void TestCallbackHandlers()
             clientSocket.Connect(ipEndpoint.Address, ipEndpoint.Port);
 
             var callbackConnection = new CallbackConnection(0, clientSocket, callbackHandlersDict);
-            Task.Run(() => callbackConnection.Run(token));
+            Task task = Task.Run(() => callbackConnection.Run(token));
 
-            using ISocketWrapper serverSocket = serverListener.Accept();
-            WriteAndReadTestData(serverSocket, callbackHandler, inputToHandler, token);
+            if (token.IsCancellationRequested)
+            {
+                task.Wait();
+                Assert.False(callbackConnection.IsRunning);
+            }
+            else
+            {
+                using ISocketWrapper serverSocket = serverListener.Accept();
+                WriteAndReadTestData(serverSocket, callbackHandler, inputToHandler);
+
+                if (callbackHandler.Throws)
+                {
+                    Assert.False(callbackConnection.IsRunning);
+                }
+                else
+                {
+                    Assert.True(callbackConnection.IsRunning);
```

Comment:
Question. When is this running `callbackConnection` stopped?

Reply:
The callbackConnection can be stopped when the JVM side closes the connection cleanly, with a "CallbackFlags.CLOSE` request, or when reading the stream the socket is closed unexpectedly, or when the cancellation token is triggerd (ie, when the callback server shutsdown).

Reply:
I mean in this function?

Reply:
It'll should stop once the GC cleans up the sockets.

Reply:
Or if we can manually trigger the cancellation token after each test.

Reply:
Yea, can you make the clean up deterministic please?

Reply:
CallbackConnections stopped.